XenAdmin/Commands/RemoveHostFromPoolCommand.cs

```diff
@@ -121,38 +120,40 @@ protected override void ExecuteCore(SelectedItemCollection selection)
             RunMultipleActions(actions, Messages.REMOVING_SERVERS_FROM_POOL, Messages.POOLCREATE_REMOVING, Messages.POOLCREATE_REMOVED, true);
         }
 
+        protected override string GetCantExecuteReasonCore(IXenObject item)
+        {
+            if (item is Host host)
+            {
+                if (host.IsMaster())
+                    return Messages.MESSAGEBOX_POOL_MASTER_REMOVE;
+
+                if (!host.IsLive())
+                    return Messages.HOST_UNREACHABLE;
+
+                if (host.resident_VMs != null && host.resident_VMs.Count > 1)
+                    return Messages.NEWPOOL_HAS_RUNNING_VMS;
+            }
+
+            return base.GetCantExecuteReasonCore(item);
+        }
+
         public static bool CanExecute(Host host)
         {
             if (host != null && host.Connection != null)
             {
                 Pool pool = Helpers.GetPool(host.Connection);
-                return pool != null && host.opaque_ref != pool.master && host.resident_VMs != null && host.resident_VMs.Count < 2 && host.IsLive();
+                return pool != null && host.opaque_ref != pool.master && host.IsLive() &&
+                       host.resident_VMs != null && host.resident_VMs.Count <= 1;
             }
             return false;
         }
 
         protected override bool CanExecuteCore(SelectedItemCollection selection)
         {
-            if (selection.AllItemsAre<Host>())
-            {
-                // all hosts must be in same pool.
-                Pool commonPool = null;
-                foreach (Host host in selection.AsXenObjects<Host>())
-                {
-                    if (!CanExecute(host))
-                    {
-                        return false;
-                    }
-                    Pool pool = Helpers.GetPool(host.Connection);
-                    if (commonPool != null && !pool.Equals(commonPool))
-                    {
-                        return false;
-                    }
-                    commonPool = pool;
-                }
-                return true;
-            }
-            return false;
+            // all selected items must be hosts and in the same pool

```

Comment:
I presume `CanExecute` itself now takes care of the "in the same pool" requirement. Its a little difficult to follow.

Reply:
The same pool requirement is the `selection.Select(s => s.Connection).Count() == 1` bit in the CanExecuteCore. If this fails either as an empty or as e multi-pool selection, it will not proceed to evaluate the second bit which calls CanExecute on each host.